File: 0_1.baekjoon/tree/python/BOJ11437.py
from sys import stdin, setrecursionlimit
from collections import defaultdict
setrecursionlimit(10 ** 6)


readline = stdin.readline
LOG = 21

N = int(readline())
tree = defaultdict(list)
parent = [[0 for _ in range(LOG)] for _ in range(N + 1)]
nodes_depth = defaultdict(int)

def solve():
    for _ in range(N - 1):
        n, m = map(int, readline().split())
        tree[n].append(m)
        tree[m].append(n)

    set_parent()

    M = int(readline())

    for _ in range(M):
        n, m = map(int, readline().split())
        print(lca(n, m))

def lca(n, m):

    if nodes_depth[n] < nodes_depth[m]:
        n, m = m, n

    for i in range(LOG - 1, -1, -1):
        if nodes_depth[n] - nodes_depth[m] >= (1 << i):
            n = parent[n][i]

    if n == m:
        return n

    for i in range(LOG - 1, -1, -1):
        
        if parent[n][i] != parent[m][i]:
            n = parent[n][i]
            m = parent[m][i]

    return parent[n][0]

def set_parent():
    dfs(1, 1)

    for d in range(1, LOG):
        for node in range(1, N + 1):
            parent[node][d] = parent[parent[node][d - 1]][d - 1]

def dfs(node, depth):
    nodes_depth[node] = depth

    for adj in tree[node]:
        if adj not in nodes_depth:
            parent[adj][0] = node
            dfs(adj, depth + 1)
solve()

File: 0_1.baekjoon/tree/java/src/BOJ11437.java
import java.util.*;
import java.io.*;

public class BOJ11437 {
    static Map<Integer, List<Integer>> tree;
    static int[] nodesDepth;
    static int[] parent;
    public static void main(String[] args) throws NumberFormatException, IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        tree = new HashMap<>();
        nodesDepth = new int[N + 1];
        parent = new int[N + 1];

        for (int i=1;i<N;i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());

            int n = Integer.parseInt(st.nextToken());
            int m = Integer.parseInt(st.nextToken());

            tree.putIfAbsent(n, new ArrayList<>());
            tree.putIfAbsent(m, new ArrayList<>());

            tree.get(n).add(m);
            tree.get(m).add(n);
        }

        dfs(1, 1);
        
        int M = Integer.parseInt(br.readLine());

        for (int i=0;i<M;i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());

            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            System.out.println(lca(a, b));
        }
    }

    private static int lca(int n, int m) {

        while (nodesDepth[n] != nodesDepth[m]) {
            if (nodesDepth[n] > nodesDepth[m]) {
                n = parent[n];
            } else {
                m = parent[m];
            }
        }

        while (n != m) {
            n = parent[n];
            m = parent[m];
        }        

        return n;
    }

    private static void dfs(int node, int depth) {
        nodesDepth[node] = depth;

        for (int adj : tree.get(node)) {
            if (nodesDepth[adj] == 0) {
                parent[adj] = node;
                dfs(adj, depth + 1);
            }
        }
    }
}
